W-18384849-Add-MUnit-processor-link-to-ACB-docs-CP (#452)
mulesoft/docs-code-builder: modules/ROOT/pages/int-test-munit.adoc

include::reuse::partial$beta-banner.adoc[tag="anypoint-code-builder"]

Create, run, and modify MUnit tests for your implementation and integration projects in Anypoint Code Builder. Import projects that contain MUnit test files. 

MUnit is a MuleSoft unit testing framework for creating and running unit tests on a Mule application, as described in xref:munit::index.adoc[MUnit Overview]. 

> With MUnit, create MUnit tests and MUnit test suites. MUnit test is an individual test case, defined by the `<munit:test>` element, which includes specific validations and assertions to make sure the functionality of a Mule application. A MUnit test suite is a container for multiple MUnit tests, organizing them within an XML file typically found in the `/src/test/munit` directory.

Reviewer: ```suggestion
With MUnit, create MUnit tests and MUnit test suites. An MUnit test is an individual test case, defined by the `<munit:test>` element, which includes specific validations and assertions to test the functionality of a Mule application. An MUnit test suite is a container for multiple MUnit tests. It organizes tests in an XML file typically found in the `/src/test/munit` directory.
```

Author: OK! Thanks!

Anypoint Code Builder provides a *Testing* panel from which you can run tests. In the *Explorer*, test suites are also located within the project directory `{munit-src}`. The IDE also shows a canvas for creating and maintaining MUnit test suits, a Testing Results panel with the status of tests, and logs that identify the status of tests. 

> For MUnit features supported by Anypoint Code Builder, see xref:acb-release-notes.adoc[Anypoint Code Builder Release Notes]. Anypoint Code Builder - MUnit Extension, part of the Anypoint Extension Pack, supports MUnit functionality in Anypoint Code Builder. For information about extensions, see xref:index.adoc#extension-pack[Anypoint Extension Pack].

Reviewer: I'm confused as to why you're linking to release notes, which change frequently. Consider removing sentence.

Author: Great catch! I'll remove it. Thanks!

[[prereqs]]
== Before You Begin

Understand the fundamentals of MUnit. Refer to the MUnit documentation for an overview and guidance with test configurations:

** xref:munit::index.adoc[]
** xref:munit::munit-test-concept.adoc[]
** xref:munit::event-processors.adoc[]
** xref:munit::munit-cookbook.adoc[]

//MUNIT WORKFLOWS AND TASKS
[[workflow]]
== MUnit Tasks and Workflows

A workflow can include one or more steps:

. <<create-tests, Creating tests and test suites>>
. <<open-test, Opening tests>>
. <<run, Running tests>>
. <<run-debug, Debugging>>
. <<fix-errors, Fixing errors>>
. <<import-test, Importing a project that contains MUnit tests>>

//
// CREATE TESTS AND TEST SUITES
[[create-tests]]
== Create MUnit Tests and Test Suites

> Configure MUnit tests and test suites using the canvas, the XML editor, or directly from a flow.

Reviewer: Love this!

Author: Thanks!

* <<create-suite>>
* <<create-test>>

[[create-suite]]
=== Create an MUnit Test Suite

Create an MUnit test suite in your project from the *Explorer*. 

. Open your project to `{munit-src}` in the *Explorer*.
. Create a suite file in the `munit` folder:
.. Right-click the `munit` folder and select *New MUnit Test Suite File*.
.. Provide a name for your suite, such as `my-suite.xml`.
+

> After the February 2025, the IDE creates a canvas with a link to build a new test with scope and flow structure options.

Reviewer: ```suggestion
A canvas appears with a link to build a new test with scope and flow structure options.
```

Author: OK! Thanks!

                 http://www.mulesoft.org/schema/mule/munit http://www.mulesoft.org/schema/mule/munit/current/mule-munit.xsd
                 http://www.mulesoft.org/schema/mule/munit-tools  
                 http://www.mulesoft.org/schema/mule/munit-tools/current/mule-munit-tools.xsd">
	
  <munit:config name="my-suite.xml" />

</mule>
----
. Select a component from the canvas to customize your test:
* With the *Behavior* components, such as 'munit:before-test', 'munit:after-test', etc., you can set the conditions for the test.
* With the *Execution* components, the '<flow-ref>'' or inline flow definition within the '<munit:execution>' block, you can specify the actions or processes to test.
* With the *Validation* components, such as 'munit:assert-that', you can check the output or state of the application after the execution phase to ensure it matches the expected results so you can confirm the application behaves as intended.
+
See xref:munit::event-processors.adoc[] for more information.
. Continue to <<create-test>>.

[NOTE]
Anypoint Code Builder requires MUnit 3.4.0 or later to create MUnit test suites from the canvas.

[[create-test]]
=== Create an MUnit Test

> Create an MUnit test in a new or existing test suite. Use the canvas, the XML Editor, or right-click a flow to start an MUnit test. For guidance with your test configuration, refer to the <<prereqs, MUnit documentation>> described in the prerequisites.

Reviewer: "right-click a flow" is confusing to me. Does the user right-click a file, a flow component, or something else?

Author: There's a component in every mule project called "flow". It's the first component in a Mule project flow. I'll make a small change to make it more easy to understand. Thanks!

==== Create a Test from a Flow
You can quickly generate the basic structure for an MUnit test designed to test a specific flow within your Mule application.

. Open your Mule project.

> . In the canvas, right-click on your flow.

Reviewer: Any component in the flow? Can the user click on whitespace between components?

Author: As commented before, it's the `flow` component they should right click on.

+
image:munit-clear-testing-results-log.png["Clearing test results from Test Results panel"]

[[observe-processors-in-your-flow]]

> == Observe Processors in Your Flow

Reviewer: "Observe" is confusing to me. My understanding is that the feature lets devs target specific processors and attributes for testing. Is there a word that conveys this better to the customer? I think "targeted testing" might be a candidate.

Author: Let's see if you like my new title 🤞

Add these processors to your MUnit test to focus on individual components of your Mule flow:

[options="header,footer",cols="1,2",]
|====
| Processor | Description
| `Mock-when` | Create a mock of an event processor in your flow to simulate the behavior of the event processor without actually executing it.
| `Spy` | Observe what happens before and after a specific event processor in your flow is executed, without altering its behavior.
|====

=== Setting Up Processors

. On your test suite, click *+*.
. Search for a processor designed to observe your flow (see <<observe-processors-in-your-flow>>).
. Select the type of processor to-be observed by clicking on the *+* icon.
//TODO: ADD IMAGE
. Configure the attributes.
//TODO: ADD IMAGE

When selecting the processor and its attributes, consider these aspects:

//TODO: REVIEW THESE CONSIDERATIONS

> * The setup configured for one processor applies to all processors of the same type in your flow.

Reviewer: I'd use "setup" or "configuration", but not both

Author: Great catch! Thanks!

+
image:munit-clear-testing-results-log.png["Clearing test results from Test Results panel"]

[[targeting-specific-processors-and-attributes-in-your-test]]
== Targeting Specific Processors and Attributes in Your Test

Add these processors to your MUnit test to focus on individual components of your Mule flow:

[options="header,footer",cols="1,2",]
|====
| Processor | Description
| `Mock-when` | Create a mock of an event processor in your flow to simulate the behavior of the event processor without actually executing it.
| `Spy` | Observe what happens before and after a specific event processor in your flow is executed, without altering its behavior.
|====

=== Setting Up Processors

. On your test suite, click *+*.
. Search for a processor designed to observe your flow (see <<observe-processors-in-your-flow>>).

> . Select the type of processor to-be observed by clicking on the *+* icon.

Reviewer: ```suggestion
. Select the type of processor to be observed by clicking on the *+* icon.
```

Author: Great catch! Thanks!
